### PR TITLE
Model3d Importer

### PR DESCRIPTION
**Pull request by pegvin**
#15

### CODE REVIEW

**Comment by csakashraj** [comment]
gud boi

**Comment by pegvin** [comment]
Will be merging it but before I'll add support for layers.

**Comment by bztsrc** [comment]
> I'll add support for layers.

What do you mean by that? Add layers to the m3d importer/exporter? Those already handle multiple layers. There's one `VOXT` chunk storing the voxel types (currently only their colors, because goxel supports cube voxels only without type names), and then one `VOXD` voxel data chunk for each layer. On import each `VOXD` chunk is loaded into a new layer.

Is this what you mean? If so, then it's already taken care of. If not, then let me know if this is something I can help with.

Cheers,
bzt

**Comment by pegvin** [comment]
@bztsrc i meant exporting layer, as you said in #15:

> It also loads back the layer's name if stored in the file (but the exported does not save any strings nor colormap into the models for simplicity).

as you said importer reads and sets layer names but exporter doesn't for simplicity, so i want to add exporting layers too.

i want to use this format as main format instead of .gox, so i need to add layers in exporter but idk how i will add materials, cameras & editor state, maybe writing it in the end? after everything?

**Comment by bztsrc** [comment]
> i want to add exporting layer names too.

Ah I see. Let me see what I can do.

> idk how i will add materials

The format supports PBR [materials](https://gitlab.com/bztsrc/model3d/-/blob/master/docs/m3d_format.md#materials) and even [inlined textures](https://gitlab.com/bztsrc/model3d/-/blob/master/docs/m3d_format.md#inlined-assets)...

> cameras & editor state

...however not these, because such info has nothing to do with the model itself (by definition it's engine-specific, and the format is engine-agnostic). That being said, you can add private chunks into the files, the only restriction is, their magic must start with a lowercase letter, for example `goxl`. You can store whatever data you want in that chunk, it won't interfere with other applications and other implementations. The SDK automatically loads additional [extra chunks](https://gitlab.com/bztsrc/model3d/-/blob/master/docs/API.md#get-application-engine-specific-data-chunks) unparsed, and even saves if you pass the `M3D_EXP_EXTRA` flag to `m3d_save()`.

But we're not using the SDK here, so we have two options:

a) I implement these:
1. I add layer names to the exporter
2. I add code to the exporter to save an arbitrary buffer
3. similarly I add code to the importer to load that buffer
4. it is then entirely up to you to fill up and parse that buffer, so you can store camera, editor state, materials etc. whatever you want.

b) I rewrite the file format plugin to use the SDK. In this case
1. strings are handled out-of-the-box
2. extra chunks handled out-of-the-box
3. one only has to fill up m3d->materials array and it will be saved / loaded automatically (but there's still a need to implement conversion between goxel material format and m3d in-memory material representation).

Which one would you prefer?

Cheers,
bzt

**Comment by bztsrc** [comment]
Ah, never mind, here's a version that saves and loads everything that goxel used to store in .gox files (strings, materials, camera, lights, etc. etc. etc.).

[m3d.c.gz](https://github.com/goxel2/goxel2/files/9264956/m3d.c.gz)

I've ended up adding 3 private chunks (detailed documentation in the source):
1. `gxel` single chunk, common goxel meta info (like lights, canvas size etc.)
2. `gxcm` camera info, it is only separated because there can be multiple cameras, so this is a multiple chunk
3. `gxlr` always stored right after the corresponding `VOXD` chunk, any additional goxel-specific layer meta info goes here

I've tested it with some simple voxel image. Without the private chunks the file size was 109 bytes. With them, about the double, 233 bytes. But it is still 90% smaller than the .gox version, which was 2238 bytes. Oh, and when I was there, I've increased the coordinate size to 32 bits, this takes an additional 12 uncompressed bytes per layer (small price I believe, this isn't included in the 109 bytes version, but is in the 233 bytes one).

There's a conceptual difference with M3D layers and goxel layers. In M3D, layers are nothing more than a group of voxels, meaning it can happily store voxels using different materials on one layer. Goxel can't do that, it requires one material per layer. To overcome, on export it always saves the same material for every voxel type used on the same layer, and on import it sets the first voxel type's material (that has a material) on the layer. This works for all .m3d files saved from goxel2, and not ideal, but does not misbehave too much when importing random .m3d files either.

The visibility is stored in the uncertainty field (0 uncertainty means always visible, any other value invisible). Everything else was added to the `gxlr` chunks, which are solely used by goxel2. Therefore strings are stored in these as-is, not using the string table.

Materials are stored as standard M3D chunks with PBR values. Here goxel is a bit odd, as it has 3 floats for the emission color under the hood, but only one float slider on the UI. This means loading back emission colors might result in small rounding errors (which you will probably never notice, because on the UI the emission slider is rounded ever more, to one decimal).

**VERY, VERY, VERY IMPORTANT** I did all the basic tests, but obviously I couldn't do a through testing for all the edge cases in this short time. I believe it's okay and should work fine, but might have some under-the-radar bugs.

Oh, running it with ASAN, the sanitizer will complain about misaligned access. The sanitizer is wrong, don't care (all the modern CPUs that goxel supports are perfectly capable of accessing an uint32_t at non 4 bytes aligned addresses). I haven't fixed this for performance reasons, because `*((uint32_t*)s)` is compiled into a single instruction (`mov` on x86, `ldr` on arm), while using `(s[3] << 24) | (s[2] << 16) | (s[1] << 8) | s[0]` would compile to a bunch of instructions, 4 `mov`s, 3 shift operators, and it also wastes a register (causing the optimizer to generate worse code for the source that follows). But if the ASAN bothers you, feel free to replace these uint32_t casted accesses to shifted bytes.

Finally, setting the `M3D_DEBUGCHUNKS` define will avoid inflation, so that you can examine the saved chunks with a hexeditor (Midnight Commander or `hexdump -C` for example) for debugging. This is irrelevant for the importer, that will load both compressed and uncompressed .m3d files transparently. (Just for the records, the uncompressed version of the aforementioned voxel image with the private chunks was 509 bytes).

Cheers,
bzt

ps: I've reconfigured my IDE, no more indentation issues ;-)

**Comment by pegvin** [comment]
@bztsrc you're a real awesome person, this is the reason i love open-source, will look into it in a sec.

**Comment by pegvin** [comment]
i wonder why you removed the "credit"

```diff
@@ -2,16 +2,70 @@
  * Model 3D .m3d format support.
  *
  * Model 3D is an application and engine netural 3D model format capable of
- * storing voxel images & much more, see https://bztsrc.gitlab.io/model3d
- * Importer & Exporter Implemented By https://github.com/bztsrc
+ * storing voxel images, see https://bztsrc.gitlab.io/model3d
 */
```

like you have done so much bruh.

**Comment by pegvin** [comment]
> Without the private chunks the file size was 109 bytes. With them, about the double, 233 bytes

this is worth it! my Castle model takes about 10 KBs in gox format and 956 bytes in m3d format.

i tested it too, and everything works, so i can finally use it as my main format but will still keep the gox format in import/export option for backwards compatibility.

now with this being done, i can finally start to work on Thumbnails, so with goxel2 there will be a program that will provide thumbnails for the model3d files.

https://docs.microsoft.com/en-us/windows/win32/shell/building-thumbnail-providers

won't be a pain, thanks alot for this!

~ Aditya

**Comment by bztsrc** [comment]
> you're a real awesome person

Thank you, but truth to be told, it is in my best interest to have a goxel release which can export M3D. Right now I have my own goxel fork patched with m3d.c, and it's a PITA for me to provide binaries. So our goals align :-) Thank you for forking goxel and doing the work its original author is too lazy to do!

> i wonder why you removed the "credit"

I haven't, I just didn't noticed you have added that, so I haven't pulled. Feel free to put it back.

> i tested it too, and everything works, so i can finally use it as my main format but will still keep the gox format in import/export option for backwards compatibility.

Good to hear! You should be able to just swap the register plugin part in gox.c and m3d.c, I've added all the extra code in gox.c into m3d.c. However you might have to remove those from gox.c when its just registered as export/import option. Most part irrelevant, but `goxel.image->path=strdup(path)` in import memleaked for me (so that line is commented out in m3d.c for now). Also gox.c clears out the canvas, I have commented that out in m3d.c because .m3d files might not have a canvas info (for the records, it is optional in .gox files too). If we clear the `goxel.image->box` on import, then canvas would get shrinked, not sure we want this to happen. Read the source, I've put in lots of comments explaining what's going on. I believe that's all (strdup(path) and memset zero canvas box) that I have copied from gox.c and then commented out.

> now with this being done, i can finally start to work on Thumbnails, so with goxel2 there will be a program that will provide thumbnails for the model3d files.

There's a define which will add preview image chunks to the model files. You can read in the first 16 bytes of a file, and if there's a `PRVW` magic at offset 8, then the model has a preview, encoded as PNG (the PNG's size right after that, at offset 12, followed by the PNG file's data from offset 16). That chunk is never compressed, so that you can fetch it easily. Be aware, that chunk is *optional*, might be missing.

This is similar in .gox files, however that has a `GOX ` chunk, a variable length `IMG ` chunk, and after that the preview in `PREV` chunk (as it happens, this chunk's format is the same as M3D's `PRVW` chunk, there's nothing new under the sun). The only difference is, `PREV` chunk's file offset isn't fixed like in it's .m3d counterpart, so you have to search for it.

I believe no other voxel formats capable of storing a preview.

(Just thinking out loud: maybe for files which lack a preview, you could read in the file and use `goxel_render_to_buf` to generate the same image you would get from decoding the PNG preview. This works for *all* supported voxel image formats, but might be a bit slow, although you could save the generated previews (and/or decoded PNGs) in a cache like GIMP does. All preview is 128 x 128 pixels, so the decoded size for each preview would be 64k. Loading and generating to a pixelbuffer should be perfectly okay and safe if you use a different, temporary `goxel` instance during thumbnail generation. It would be better if the import function got the goxel instance, but sadly not. So you'll have to copy the goxel struct, clear it, call import, free goxel and copy original goxel struct back. For some reason even `goxel.image` is used directly in gox.c, so I have used that instead of the parameter in m3d.c too. I don't know why isn't it `image` (I think that should work), but maybe there's a valid reason why directly referencing is important, so I just did the way gox.c does.)

Cheers,
bzt

**Comment by pegvin** [comment]
@bztsrc yea yea, so basically this is what i was planning, my thumbnail provider will read the file and provide the thumbnail with 2 ways

1. if no preview stuff found render it using the goxel (obviously not the path tracer but the default renderer)
2. or i will just use the preview content provided in the file

> might be a bit slow, although you could save the generated previews

yea so basically, this thumbnail provider, after it provides the thumbnail the operating system automatically caches it and only calls the thumbnail provider again when the file is changed.

> For some reason even goxel.image is used directly in gox.c, so I have used that instead of the parameter in m3d.c too. I don't know why isn't it image (I think that should work)

so you get the `image_t*` only in `FILE_FORMAT_REGISTER` macro and maybe some others.

**Comment by bztsrc** [comment]
> so you get the image_t* only in FILE_FORMAT_REGISTER macro and maybe some others.

Yes, the import / export functions receive that, but for some reason gox.c doesn't use the parameter, instead it references `goxel.image` directly. I did the same to be on the safe side. Also there are other goxel fields (like `goxel.rend.light` which are not part of the image, yet saved and loaded, so the parameter really should been the entire goxel struct, not just `goxel.image`.

Cheers,
bzt

**Comment by bztsrc** [comment]
If it's not to much to ask, could you please add *Model3D, [MIT](https://bztsrc.gitlab.io/model3d/)* to the readme.md in the "Credits & Licenses" table?

Thanks,
bzt

**Comment by pegvin** [comment]
yea yea don't worry, i didn't forget that, i'm a little busy with some other stuff.
